include/RAJA/pattern/nested/internal.hpp

```diff
@@ -93,8 +105,314 @@ using index_tuple_from_segments =
     typename camp::apply_l<camp::lambda<camp::tuple>,
                            value_type_list_from_segments<Segments> >::type;
 
+
+
+
+template <typename Policy>
+struct StatementExecutor{};
+
+
+template <typename PolicyType, typename SegmentTuple,typename ... Bodies>
+struct LoopData {
+
+  using Self = LoopData<PolicyType, SegmentTuple, Bodies...>;
+
+  using offset_tuple_t = difftype_tuple_from_segments<typename SegmentTuple::TList>;
+
+  using index_tuple_t = index_tuple_from_segments<typename SegmentTuple::TList>;
+
+  using policy_t = PolicyType;
+
+  using segment_tuple_t = camp::decay<SegmentTuple>;
+  SegmentTuple segment_tuple;
+
+  using BodiesTuple = camp::tuple<Bodies...> ;
+  const BodiesTuple bodies;
+  offset_tuple_t offset_tuple;
+
+  int shmem_window_start[segment_tuple_t::TList::size];
+
+  RAJA_INLINE
+  LoopData(SegmentTuple const &s, Bodies const & ... b)
+      : segment_tuple{s}, bodies{b...}
+  {
+    for(size_t i = 0;i < segment_tuple_t::TList::size; ++ i){
+      shmem_window_start[i] = 0;
+    }
+    assign_begin_all();
+  }
+
+  template <typename PolicyType0, typename SegmentTuple0, typename ... Bodies0>
+  RAJA_INLINE
+  RAJA_HOST_DEVICE
+  constexpr
+  LoopData(LoopData<PolicyType0, SegmentTuple0, Bodies0...> &c)
+      : segment_tuple{c.segment_tuple}, bodies{c.bodies}, offset_tuple{c.offset_tuple}
+  {
+  }
+
+  template <camp::idx_t Idx, typename IndexT>
+  RAJA_HOST_DEVICE
+  RAJA_INLINE
+  void assign_offset(IndexT const &i)
+  {
+    camp::get<Idx>(offset_tuple) = i;
+  }
+
+
+  template <camp::idx_t Idx>
+  RAJA_HOST_DEVICE
+  RAJA_INLINE
+  int assign_begin()
+  {
+    camp::get<Idx>(offset_tuple) = 0;
+    return 0;
+  }
+
+  template <camp::idx_t ... Idx>
+  RAJA_HOST_DEVICE
+  RAJA_INLINE
+  void assign_begin_all_expanded(camp::idx_seq<Idx...> const &)
+  {
+    VarOps::ignore_args( assign_begin<Idx>()... );
+
+  }
+
+  RAJA_HOST_DEVICE
+  RAJA_INLINE
+  void assign_begin_all()
+  {
+    assign_begin_all_expanded(camp::make_idx_seq_t<offset_tuple_t::TList::size>{});
+  }
+
+
+
+  template <camp::idx_t ... Idx>
+  RAJA_HOST_DEVICE
+  RAJA_INLINE
+  index_tuple_t get_begin_index_tuple_expanded(camp::idx_seq<Idx...> const &) const
+  {
+    return camp::make_tuple( (*camp::get<Idx>(segment_tuple).begin())... );
+  }
+
+  RAJA_HOST_DEVICE
+  RAJA_INLINE
+  index_tuple_t get_begin_index_tuple() const
+  {
+    return get_begin_index_tuple_expanded(camp::make_idx_seq_t<offset_tuple_t::TList::size>{});
+  }
+
+};
+
+
+
+
+RAJA_SUPPRESS_HD_WARN
+template <camp::idx_t LoopIndex, camp::idx_t ... Idx, typename Data>
+RAJA_HOST_DEVICE
+RAJA_INLINE
+void invoke_lambda_expanded(camp::idx_seq<Idx...> const &, Data &data)
+{
+  camp::get<LoopIndex>(data.bodies)(
+            (camp::get<Idx>(data.segment_tuple).begin()[camp::get<Idx>(data.offset_tuple)] )...
```

Comment:
I don't like it either.  It would be nice to have another mechanism for getting non-segment things into the lambda's, like a tuple of parameters in addition to the tuple of segments.